### PR TITLE
Fix half precision base support

### PR DESCRIPTION
**Pull request by janciesko**
Excludes Maxwell{50,52} and Keppler{37,32,35,30}.

### CODE REVIEW

**Review by crtrott — APPROVED** [review]

File: core/src/Cuda/Kokkos_Cuda_Half.hpp [hunk 1]
```diff
@@ -47,12 +47,15 @@
 
 #include <Kokkos_Macros.hpp>
 #ifdef KOKKOS_ENABLE_CUDA
+#if !(defined(KOKKOS_COMPILER_CLANG) && KOKKOS_COMPILER_CLANG < 900) &&  \
+    !(defined(KOKKOS_ARCH_KEPLER30) || defined(KOKKOS_ARCH_KEPLER32) ||  \
+      defined(KOKKOS_ARCH_KEPLER35) || defined(KOKKOS_ARCH_KEPLER37)) && \
+    !(defined(KOKKOS_ARCH_MAXWELL50) || defined(KOKKOS_ARCH_MAXWELL52))
```
Comment (dalg24) [inline core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: Please confirm that `KOKKOS_ARCH_MAXWELL53` was deliberately omitted

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: Oh I didn't know 52 was a thing but yeah you are right that one is missing too.

Comment (dalg24) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: Then use `KOKKOS_ARCH_KEPLER` and `KOKKOS_ARCH_MAXWELL`

Comment (DavidPoliakoff) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: This is enough lines of preprocesor that I'd like a comment stating when this thing shouldn't be enabled

Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: Confirming that KOKKOS_ARCH_MAXWELL53 work.

Comment (janciesko) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: #ifdef KOKKOS_ENABLE_CUDA
#if !(defined(KOKKOS_COMPILER_CLANG)) && defined(KOKKOS_ARCH_MAXWELL53) || \
    defined(Kokkos_ARCH_PASCAL60) || defined(Kokkos_ARCH_PASCAL61) ||      \
    defined(Kokkos_ARCH_VOLTA70) || defined(Kokkos_ARCH_VOLTA73) ||        \
    defined(Kokkos_ARCH_MAXWELL53)
#include <cuda_fp16.h>
@DavidPoliakoff

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: I don't want the counter comment right now: 1) its gonna be as confusing as the exclusion one, and 2) we need to update it everytime a new GPU comes out.

Comment (crtrott) [reply core/src/Cuda/Kokkos_Cuda_Half.hpp:53]: but basically it seems it works for everything newer than 52.

**Pull request merged**